core/store/src/db.rs

```diff
@@ -0,0 +1,187 @@
+use parking_lot::RwLock;
+use rocksdb::{
+    ColumnFamily, ColumnFamilyDescriptor, DBCompactionStyle, DBIterator, IteratorMode, Options,
+    ReadOptions, WriteBatch, DB,
+};
+use std::collections::HashMap;
+use std::io;
+
+fn other_io_err<E>(e: E) -> io::Error
+where
+    E: Into<Box<dyn std::error::Error + Send + Sync>>,
+{
+    io::Error::new(io::ErrorKind::Other, e)
+}
+
+pub struct DBTransaction {
+    pub ops: Vec<DBOp>,
+}
+
+pub enum DBOp {
+    Insert { col: usize, key: Vec<u8>, value: Vec<u8> },
+    Delete { col: usize, key: Vec<u8> },
+}
+
+impl DBTransaction {
+    pub fn put<K: AsRef<[u8]>, V: AsRef<[u8]>>(&mut self, col: usize, key: K, value: V) {
+        self.ops.push(DBOp::Insert {
+            col,
+            key: key.as_ref().to_owned(),
+            value: value.as_ref().to_owned(),
+        });
+    }
+
+    pub fn delete<K: AsRef<[u8]>>(&mut self, col: usize, key: K) {
+        self.ops.push(DBOp::Delete { col, key: key.as_ref().to_owned() });
+    }
+}
+
+pub struct RocksDB {
+    db: RwLock<Option<DB>>,
+    cfs: Vec<*const ColumnFamily>,
+    read_options: ReadOptions,
+}
+
+unsafe impl Send for RocksDB {}
+unsafe impl Sync for RocksDB {}
+
+pub struct TestDB {
+    db: RwLock<Option<Vec<HashMap<Vec<u8>, Vec<u8>>>>>,
+}
+
+pub trait Database: Sync + Send {
+    fn transaction(&self) -> DBTransaction {
+        DBTransaction { ops: Vec::new() }
+    }
+    fn get(&self, col: usize, key: &[u8]) -> Result<Option<Vec<u8>>, io::Error>;
+    fn iter<'a>(&'a self, column: usize) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a>;
+    fn write(&self, batch: DBTransaction) -> io::Result<()>;
+}
+
+impl RocksDB {
+    fn iterate(&self, col: usize) -> Option<DBIterator<'_>> {
+        match *self.db.read() {
+            Some(ref db) => unsafe {
+                let cf_handle = &*self.cfs[col];
+                let iterator =
+                    db.iterator_cf_opt(cf_handle, &self.read_options, IteratorMode::Start).unwrap();
+                Some(iterator)
+            },
+            _ => None,
+        }
+    }
+}
+
+impl Database for RocksDB {
+    fn get(&self, col: usize, key: &[u8]) -> Result<Option<Vec<u8>>, io::Error> {
+        match *self.db.read() {
+            Some(ref db) => unsafe {
+                Ok(db.get_cf_opt(&*self.cfs[col], key, &self.read_options).map_err(other_io_err)?)
+            },
+            None => Ok(None),
+        }
+    }
+
+    fn iter<'a>(&'a self, col: usize) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
```

Comment:
It does looks suspicious. and iterator_cf_opt/iterator_cf doesn't have that 'a:'b. iterator_opt/iterator have it. They should be all same since iterator_opt/iterator is iterating "default" column family, just a special case of iterator_cf_opt/iterator_cf

Reply:
IMO this happens to correct because we never close db, so db is ~= 'static. `'b: 'a` makes lifetime of iterator >= lifetime of db ~= 'static. so iterator is also static